File: src/rust/gnomesort.rs
fn main(){
  println!("{:?}", gnomesort(vec![1,3,5,2]));
}
fn gnomesort(mut slice: Vec<i32>) -> Vec<i32> {
  let mut pivot: i32 = 0;
  let size: i32 = slice.len() as i32;

  while pivot < size - 1 {
    if slice[pivot as usize] > slice[pivot as usize + 1] {
      let temp: i32 = slice[pivot as usize];
      slice[pivot as usize] = slice[pivot as usize + 1];
      slice[pivot as usize + 1] = temp;
      if pivot > 0 {
        pivot -= 2;
      }
    }
    pivot += 1;
  }
  slice
}

File: src/go/gnome_sort.go
package main

import "fmt"

func GnomeSort(slice []int) {
	pivot := 0
	size := len(slice)

	for pivot < size-1 {
		if slice[pivot] > slice[pivot+1] {
			slice[pivot], slice[pivot+1] = slice[pivot+1], slice[pivot]
			if pivot > 0 {
				pivot -= 2
			}
		}
		pivot++
	}
}

func main() {
  slice := []int{2, 5, 1, 7, 4, 9, 8, 6, 3}
	fmt.Println("Slice:", slice)
	GnomeSort(slice)
	fmt.Println("GnomeSort:", slice)
}
